In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
ls drive/MyDrive/CaseStudy/

'AirBnB_Prediction_Challenge_Case Study and guidelines.docx'   test.xlsx
'case study.zip'                                               train.csv
'case study.zip (Unzipped Files)'/


In [ ]:
working_directory = "drive/MyDrive/CaseStudy/"

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
Train_Dataset = pd.read_csv(working_directory + 'train.csv')

In [ ]:
Train_Dataset.shape

(49999, 30)

In [ ]:
Train_Dataset.isnull().sum()

id                            0
property_type                 0
room_type                     0
amenities                     0
accommodates                  0
bathrooms                   146
bed_type                      0
cancellation_policy           0
cleaning_fee                  0
city                          0
description                   0
first_review              10711
host_has_profile_pic        127
host_identity_verified      127
host_response_rate        12338
host_since                  127
instant_bookable              0
last_review               10683
latitude                      0
longitude                     0
name                          0
neighbourhood              4617
number_of_reviews             0
review_scores_rating      11301
thumbnail_url              5595
zipcode                     640
bedrooms                     62
beds                         91
log_price                     0
                          49999
dtype: int64

In [ ]:
Train_Dataset.columns

Index(['id', 'property_type', 'room_type', 'amenities', 'accommodates',
       'bathrooms', 'bed_type', 'cancellation_policy', 'cleaning_fee', 'city',
       'description', 'first_review', 'host_has_profile_pic',
       'host_identity_verified', 'host_response_rate', 'host_since',
       'instant_bookable', 'last_review', 'latitude', 'longitude', 'name',
       'neighbourhood', 'number_of_reviews', 'review_scores_rating',
       'thumbnail_url', 'zipcode', 'bedrooms', 'beds', 'log_price', ' '],
      dtype='object')

In [ ]:
Train_Dataset.head()

id property_type        room_type  ... beds  log_price    
0   6901257     Apartment  Entire home/apt  ...  1.0   5.010635 NaN
1   6304928     Apartment  Entire home/apt  ...  3.0   5.129899 NaN
2   7919400     Apartment  Entire home/apt  ...  3.0   4.976734 NaN
3  13418779         House  Entire home/apt  ...  2.0   6.620073 NaN
4   3808709     Apartment  Entire home/apt  ...  1.0   4.744932 NaN

[5 rows x 30 columns]

In [ ]:
Train_Dataset.dtypes

id                          int64
property_type              object
room_type                  object
amenities                  object
accommodates                int64
bathrooms                 float64
bed_type                   object
cancellation_policy        object
cleaning_fee                 bool
city                       object
description                object
first_review               object
host_has_profile_pic       object
host_identity_verified     object
host_response_rate         object
host_since                 object
instant_bookable           object
last_review                object
latitude                  float64
longitude                 float64
name                       object
neighbourhood              object
number_of_reviews           int64
review_scores_rating      float64
thumbnail_url              object
zipcode                    object
bedrooms                  float64
beds                      float64
log_price                 float64
              

In [ ]:
cleaned_train_dataset = Train_Dataset.drop(['id',
                                            'zipcode',
                                            'cancellation_policy',
                                            'description','first_review',
                                            'host_has_profile_pic',
                                            'host_identity_verified',
                                            'host_response_rate',
                                            'host_since',
                                            'instant_bookable',
                                            'last_review', 
                                            'latitude', 
                                            'longitude', 
                                            'name',
                                            'neighbourhood',
                                            'thumbnail_url',
                                            ' '],axis = 'columns',inplace = False)

In [ ]:
cleaned_train_dataset.head()

property_type        room_type  ... beds  log_price
0     Apartment  Entire home/apt  ...  1.0   5.010635
1     Apartment  Entire home/apt  ...  3.0   5.129899
2     Apartment  Entire home/apt  ...  3.0   4.976734
3         House  Entire home/apt  ...  2.0   6.620073
4     Apartment  Entire home/apt  ...  1.0   4.744932

[5 rows x 13 columns]

property_type: Categorial,
  room_type: Categorical,
  amenities: Categorical list,
  accommodates: numerical,
  bathrooms: numerical(float),
  bed_type: categorical,
  cleaning_fee: Boolean,
  review_scores_rating: numberical float,
  number_of_reviews: numerical,
  bedrooms: numerical,
  beds:numerical,
  log_price: numerical,

In [ ]:
cleaned_train_dataset.shape

(49999, 13)

### Missing Values Handling

In [ ]:
cleaned_train_dataset.isnull().sum()

property_type               0
room_type                   0
amenities                   0
accommodates                0
bathrooms                 146
bed_type                    0
cleaning_fee                0
city                        0
number_of_reviews           0
review_scores_rating    11301
bedrooms                   62
beds                       91
log_price                   0
dtype: int64

In [ ]:
sum(cleaned_train_dataset.apply(lambda x: sum(x.isnull().values), axis = 1)>0)

11477

Above we can observe that summation of all missing values is 11477, review_score_rating has highest missing values among all.

In [ ]:
# Replacing missing values with 0 for review_scores_rating column
cleaned_train_dataset['review_scores_rating'] = cleaned_train_dataset['review_scores_rating'].fillna(0)

In [ ]:
total_missing_values = sum(cleaned_train_dataset.apply(lambda x: sum(x.isnull().values), axis = 1)>0)

In [ ]:
# Percentage of missing values in the given dataset
(total_missing_values/len(cleaned_train_dataset)) * 100

0.47800956019120383

As missing values are very low i.e., 0.47%, will do not have much affect in the results

In [ ]:
cleaned_train_dataset.dropna(inplace=True)

In [ ]:
cleaned_train_dataset.isnull().sum()

property_type           0
room_type               0
amenities               0
accommodates            0
bathrooms               0
bed_type                0
cleaning_fee            0
city                    0
number_of_reviews       0
review_scores_rating    0
bedrooms                0
beds                    0
log_price               0
dtype: int64

In [ ]:
cleaned_train_dataset.dtypes

property_type            object
room_type                object
amenities                object
accommodates              int64
bathrooms               float64
bed_type                 object
cleaning_fee               bool
city                     object
number_of_reviews         int64
review_scores_rating    float64
bedrooms                float64
beds                    float64
log_price               float64
dtype: object

### Label Encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
cleaning_fee_label_encoder = LabelEncoder()
cleaning_fee_label_encoder.fit(cleaned_train_dataset['cleaning_fee'])

LabelEncoder()

In [ ]:
cleaning_fee_label_encoder.classes_

array([False,  True])

In [ ]:
cleaned_train_dataset['cleaning_fee'].dtype

dtype('bool')

In [ ]:
cleaned_train_dataset['cleaning_fee'] =  cleaning_fee_label_encoder.transform(cleaned_train_dataset['cleaning_fee'])

In [ ]:
cleaned_train_dataset['cleaning_fee'].dtype

dtype('int64')

In [ ]:
cleaned_train_dataset['cleaning_fee']

0        1
1        1
2        1
3        1
4        1
        ..
49994    0
49995    1
49996    1
49997    1
49998    0
Name: cleaning_fee, Length: 49760, dtype: int64

### Handling amenties feature

In [ ]:
for i in cleaned_train_dataset['amenities'][:10]:
  print(i)

{"Wireless Internet","Air conditioning",Kitchen,Heating,"Family/kid friendly",Essentials,"Hair dryer",Iron,"translation missing: en.hosting_amenity_50"}
{"Wireless Internet","Air conditioning",Kitchen,Heating,"Family/kid friendly",Washer,Dryer,"Smoke detector","Fire extinguisher",Essentials,Shampoo,Hangers,"Hair dryer",Iron,"translation missing: en.hosting_amenity_50"}
{TV,"Cable TV","Wireless Internet","Air conditioning",Kitchen,Breakfast,"Buzzer/wireless intercom",Heating,"Family/kid friendly","Smoke detector","Carbon monoxide detector","Fire extinguisher",Essentials,Shampoo,Hangers,"Hair dryer",Iron,"Laptop friendly workspace","translation missing: en.hosting_amenity_50"}
{TV,"Cable TV",Internet,"Wireless Internet",Kitchen,"Indoor fireplace","Buzzer/wireless intercom",Heating,Washer,Dryer,"Smoke detector","Carbon monoxide detector","First aid kit","Fire extinguisher",Essentials}
{TV,Internet,"Wireless Internet","Air conditioning",Kitchen,"Elevator in building",Heating,"Smoke detecto

In [ ]:
def process_amenities(text):
    words = text.replace("{","").replace("}","").replace('"',"").replace(' ',"").split(",")
    if "translation missing:en.hosting_amenity_50" in words:
        return " ".join(words[:-1])
    else:
        return " ".join(words)

In [ ]:
cleaned_train_dataset['amenities'] = cleaned_train_dataset['amenities'].apply(process_amenities)

In [ ]:
cleaned_train_dataset['amenities']

0        WirelessInternet Airconditioning Kitchen Heati...
1        WirelessInternet Airconditioning Kitchen Heati...
2        TV CableTV WirelessInternet Airconditioning Ki...
3        TV CableTV Internet WirelessInternet Kitchen I...
4        TV Internet WirelessInternet Airconditioning K...
                               ...                        
49994    TV WirelessInternet Airconditioning Kitchen El...
49995    TV WirelessInternet Airconditioning Kitchen Br...
49996    TV Internet WirelessInternet Airconditioning K...
49997    WirelessInternet Kitchen Petsallowed Buzzer/wi...
49998    Internet WirelessInternet Airconditioning Kitc...
Name: amenities, Length: 49760, dtype: object

## Applying BAG OF WORD(BOW) on amenities feature

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
vectorizer = CountVectorizer()

In [ ]:
df = pd.DataFrame(vectorizer.fit_transform(cleaned_train_dataset['amenities']).toarray())

In [ ]:
vectorizer.get_feature_names()

['24',
 'accessible',
 'airconditioning',
 'airpurifier',
 'babybath',
 'babymonitor',
 'babysitterrecommendations',
 'bathtowel',
 'bathtub',
 'bathtubwithshowerchair',
 'bbqgrill',
 'beachessentials',
 'beachfront',
 'bedlinens',
 'bodysoap',
 'breakfast',
 'buzzer',
 'cabletv',
 'carbonmonoxidedetector',
 'cat',
 'changingtable',
 'children',
 'cleaningbeforecheckout',
 'coffeemaker',
 'cookingbasics',
 'crib',
 'darkeningshades',
 'disabledparkingspot',
 'dishesandsilverware',
 'dishwasher',
 'dog',
 'doorman',
 'doormanentry',
 'dryer',
 'elevator',
 'elevatorinbuilding',
 'en',
 'essentials',
 'ethernetconnection',
 'evcharger',
 'extrapillowsandblankets',
 'family',
 'fireextinguisher',
 'fireplaceguards',
 'firmmatress',
 'firmmattress',
 'firstaidkit',
 'fixedgrabbarsforshower',
 'flat',
 'flatsmoothpathwaytofrontdoor',
 'freeaccess',
 'freeparkingonpremises',
 'freeparkingonstreet',
 'gameconsole',
 'gardenorbackyard',
 'grab',
 'groundflooraccess',
 'gym',
 'hairdryer',
 'ha

In [ ]:
import pickle

def save_model(model, filename="model.sav"):
    pickle.dump(model, open(filename, 'wb'))
 

In [ ]:
save_model(vectorizer, "bow_vectorizer.sav")

In [ ]:
df

0    1    2    3    4    5    6    ...  138  139  140  141  142  143  144
0        0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1
1        0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1
2        0    0    1    0    0    0    0  ...    0    0    0    0    0    1    1
3        0    0    0    0    0    0    0  ...    0    0    0    0    0    1    1
4        0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
49755    0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1
49756    0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1
49757    1    0    1    0    0    0    0  ...    0    0    0    0    0    1    1
49758    0    0    0    0    0    0    0  ...    0    0    0    0    0    1    1
49759    0    0    1    0    0    0    0  ...    0    0    0    0    0    0    1

[49760 rows x 145 columns]

In [ ]:
df.shape

(49760, 145)

In [ ]:
cleaned_train_dataset.shape

(49760, 13)

In [ ]:
df.to_numpy().shape

(49760, 145)

In [ ]:
cleaned_train_dataset.to_numpy().shape

(49760, 13)

In [ ]:
cleaned_train_dataset = cleaned_train_dataset.drop(['amenities',],axis = 'columns',inplace = False)

In [ ]:
amenities_cleaned_train_dataset = pd.DataFrame(np.concatenate((cleaned_train_dataset.to_numpy(), df.to_numpy()),axis=1))

In [ ]:
amenities_cleaned_train_dataset

0                1   2   3         4   5    ... 151 152 153 154 155 156
0      Apartment  Entire home/apt   3   1  Real Bed   1  ...   0   0   0   0   0   1
1      Apartment  Entire home/apt   7   1  Real Bed   1  ...   0   0   0   0   0   1
2      Apartment  Entire home/apt   5   1  Real Bed   1  ...   0   0   0   0   1   1
3          House  Entire home/apt   4   1  Real Bed   1  ...   0   0   0   0   1   1
4      Apartment  Entire home/apt   2   1  Real Bed   1  ...   0   0   0   0   0   1
...          ...              ...  ..  ..       ...  ..  ...  ..  ..  ..  ..  ..  ..
49755  Apartment     Private room   1   1  Real Bed   0  ...   0   0   0   0   0   1
49756  Apartment  Entire home/apt   4   1  Real Bed   1  ...   0   0   0   0   0   1
49757  Apartment  Entire home/apt   5   1  Real Bed   1  ...   0   0   0   0   1   1
49758  Apartment     Private room   2   1  Real Bed   1  ...   0   0   0   0   1   1
49759      House     Private room   2   1  Real Bed   0  ...   0   0   0   0   0   1

[49760 rows x 157 columns]

In [ ]:
amenities_cleaned_train_dataset.dtypes

0      object
1      object
2      object
3      object
4      object
        ...  
152    object
153    object
154    object
155    object
156    object
Length: 157, dtype: object

In [ ]:
amenities_cleaned_train_dataset.columns

RangeIndex(start=0, stop=157, step=1)

### Divide Features and Label

In [ ]:
X = amenities_cleaned_train_dataset.iloc[:, amenities_cleaned_train_dataset.columns != 11].values
y = amenities_cleaned_train_dataset.iloc[:, 11].values


In [ ]:
# For separate algo
X = amenities_cleaned_train_dataset.iloc[:, :].values



In [ ]:
X.shape


(49760, 156)

In [ ]:
y.shape

(49760,)

In [ ]:
y

array([5.010635294, 5.129898715, 4.9767337419999995, ..., 5.068904202,
       4.234106505, 3.871201011], dtype=object)

### One Hot Encoding of categorical variables

In [ ]:
X

array([['Apartment', 'Entire home/apt', 3, ..., 0, 0, 1],
       ['Apartment', 'Entire home/apt', 7, ..., 0, 0, 1],
       ['Apartment', 'Entire home/apt', 5, ..., 0, 1, 1],
       ...,
       ['Apartment', 'Entire home/apt', 5, ..., 0, 1, 1],
       ['Apartment', 'Private room', 2, ..., 0, 1, 1],
       ['House', 'Private room', 2, ..., 0, 0, 1]], dtype=object)

#### property_type, room_type,bed_type,city

In [ ]:
print(len(cleaned_train_dataset['property_type'].value_counts()))
cleaned_train_dataset['property_type'].value_counts()

33


Apartment             32955
House                 11083
Condominium            1756
Townhouse              1144
Loft                    831
Other                   389
Guesthouse              324
Bed & Breakfast         316
Bungalow                263
Villa                   119
Dorm                     92
Guest suite              83
Camper/RV                62
Cabin                    49
In-law                   49
Hostel                   48
Timeshare                46
Boutique hotel           44
Boat                     36
Serviced apartment       16
Tent                     12
Castle                    8
Yurt                      7
Vacation home             7
Treehouse                 6
Hut                       5
Chalet                    3
Earth House               2
Train                     1
Cave                      1
Casa particular           1
Tipi                      1
Lighthouse                1
Name: property_type, dtype: int64

In [ ]:
print(len(cleaned_train_dataset['room_type'].value_counts()))
cleaned_train_dataset['room_type'].value_counts()

3


Entire home/apt    27811
Private room       20528
Shared room         1421
Name: room_type, dtype: int64

In [ ]:
print(len(cleaned_train_dataset['bed_type'].value_counts()))
cleaned_train_dataset['bed_type'].value_counts()

5


Real Bed         48347
Futon              513
Pull-out Sofa      400
Airbed             320
Couch              180
Name: bed_type, dtype: int64

In [ ]:
print(len(cleaned_train_dataset['city'].value_counts()))
cleaned_train_dataset['city'].value_counts()

6


NYC        21685
LA         15114
SF          4338
DC          3789
Chicago     2496
Boston      2338
Name: city, dtype: int64

In [ ]:
X[:,0]

array(['Apartment', 'Apartment', 'Apartment', ..., 'Apartment',
       'Apartment', 'House'], dtype=object)

In [ ]:
X[:,0]

array(['Apartment', 'Apartment', 'Apartment', ..., 'Apartment',
       'Apartment', 'House'], dtype=object)

In [ ]:
X[:,1]

array(['Entire home/apt', 'Entire home/apt', 'Entire home/apt', ...,
       'Entire home/apt', 'Private room', 'Private room'], dtype=object)

In [ ]:
X[:,1]

array(['Entire home/apt', 'Entire home/apt', 'Entire home/apt', ...,
       'Entire home/apt', 'Private room', 'Private room'], dtype=object)

In [ ]:
X[:,4]

array(['Real Bed', 'Real Bed', 'Real Bed', ..., 'Real Bed', 'Real Bed',
       'Real Bed'], dtype=object)

In [ ]:
X[:,4]

array(['Real Bed', 'Real Bed', 'Real Bed', ..., 'Real Bed', 'Real Bed',
       'Real Bed'], dtype=object)

In [ ]:
X[:,6]

array(['NYC', 'NYC', 'NYC', ..., 'NYC', 'NYC', 'NYC'], dtype=object)

In [ ]:
X[:,6]

array(['NYC', 'NYC', 'NYC', ..., 'NYC', 'NYC', 'NYC'], dtype=object)

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

In [ ]:
X.shape

(49760, 157)

In [ ]:
X

array([['Apartment', 'Entire home/apt', 3, ..., 0, 0, 1],
       ['Apartment', 'Entire home/apt', 7, ..., 0, 0, 1],
       ['Apartment', 'Entire home/apt', 5, ..., 0, 1, 1],
       ...,
       ['Apartment', 'Entire home/apt', 5, ..., 0, 1, 1],
       ['Apartment', 'Private room', 2, ..., 0, 1, 1],
       ['House', 'Private room', 2, ..., 0, 0, 1]], dtype=object)

In [ ]:
ct = ColumnTransformer([("encoder",
                         OneHotEncoder(),
                         [0,1,4,6]
                         )],
                        remainder='passthrough')

In [ ]:
X[0]

array(['Apartment', 'Entire home/apt', 3, 1.0, 'Real Bed', 1, 'NYC', 2,
       100.0, 1.0, 1.0, 5.010635294, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1], dtype=object)

In [ ]:
y[0]

5.010635294

In [ ]:
X = ct.fit_transform(X)

In [ ]:
X[0]

array([1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
       0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
       0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
       0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 3, 1.0, 1, 2, 100.0, 1.0,
       1.0, 5.010635294, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
      dtype=object)

In [ ]:
X

array([[1.0, 0.0, 0.0, ..., 0, 0, 1],
       [1.0, 0.0, 0.0, ..., 0, 0, 1],
       [1.0, 0.0, 0.0, ..., 0, 1, 1],
       ...,
       [1.0, 0.0, 0.0, ..., 0, 1, 1],
       [1.0, 0.0, 0.0, ..., 0, 1, 1],
       [0.0, 0.0, 0.0, ..., 0, 0, 1]], dtype=object)

In [ ]:
X.shape

(49760, 200)

In [ ]:
y

array([5.010635294, 5.129898715, 4.9767337419999995, ..., 5.068904202,
       4.234106505, 3.871201011], dtype=object)

Now all features in numberical form.

In [ ]:
pd.DataFrame(X).to_csv(working_directory+'Cleaned_Train.csv',index=False)

In [ ]:
# Deleting unnecessary variables to save memory
del df
del cleaned_train_dataset
del amenities_cleaned_train_dataset
del Train_Dataset